### PR TITLE
Adding test to verify if 'Unregister system' works correctly for RHAI.

### PR DESCRIPTION
**Pull request by anarang**

### CODE REVIEW

File: tests/foreman/rhai/test_rhai.py [hunk 1]
```diff
@@ -101,3 +102,35 @@ def test_org_selection_for_rhai(self):
...
+        self.rhai.register_client_to_rhai(
+            self.ak_name,
+            self.org_label,
+        )
+        with Session(self.browser) as session:
+            session.nav.go_to_select_org(self.org_name)
+            Navigator(self.browser).go_to_insights_systems()
+            # Click on the unregister icon 'X' in the table against the
+            # registered system listed.
+            session.nav.wait_until_element(
```
Comment (oshtaier) [inline tests/foreman/rhai/test_rhai.py]: Please, use click() method according to https://github.com/SatelliteQE/robottelo/pull/2556

File: tests/foreman/rhai/test_rhai.py [hunk 2]
```diff
@@ -101,3 +102,35 @@ def test_org_selection_for_rhai(self):
...
+                locators['insights.unregister_system']).click()
+            # Confirm selection for clicking on 'Yes' to unregister the system
+            session.nav.wait_until_element(
```
Comment (oshtaier) [inline tests/foreman/rhai/test_rhai.py]: Same here

**Comment by elyezer** [comment]
I think we need a better approach to register a client to RHAI. I recommend moving the register to RHAI to VirtualMachine class. This way will be possible to run

```python
with VirtualMachine(...) as vm:
    vm.register_to_rhai(...)
    # do some assertions...
```

This way is any AssertionError is raised, the client machine will not be destroyed. And by using the with statement that will be avoided.

Another approach is to add the VM creation in the setUp and tearDown methods, this way the destroy can live in the tearDown and will be executed.

**Comment by anarang** [comment]
@elyezer initially I was using the context manager as the VM getting destroyed did not affect the test case but I removed it since I wanted to access the VM later after doing an 'unregister' from the UI. I also, thought of adding the VM to the setUp class but that would not work as then only one VM would be created and test cases will not be independent of each other, so that is ruled out too. I completely agree that if an Exception is raised the VM will not be destroyed, so is there a way  you can think of other than the above two (or lastly, try catch exception) to handle this scenario?

**Comment by oshtaier** [comment]
If you want some statements to be executed not for whole class, but per each test case, please use `setUp`, not the `setUpClass` method

**Comment by anarang** [comment]
@oshtaier as discussed on irc, we cannot have setup method in this scenario where I may require to setup a VM on selected test cases and destroy them selectively.
@elyezer do you have an opinion how we should go about such a scenario.

File: tests/foreman/rhai/test_rhai.py (hunk 1 above)
Comment (anarang) [inline tests/foreman/rhai/test_rhai.py]: Done. Used click() as suggested.

File: tests/foreman/rhai/test_rhai.py (hunk 2 above)
Comment (anarang) [inline tests/foreman/rhai/test_rhai.py]: Done

**Comment by elyezer** [comment]
@anarang I can't imagine the entire scenario. Do you already have the test cases in place? Also there are VM machines being left running on ginger and it was not happening before the RHAI tests. The only concern here is that all machines must be destroyed.

I can't offer a suggestion right now as I still understanding the entire RHAI thing.

**Comment by anarang** [comment]
@elyezer apologies for the VM machines being left running, the test case which requires a VM to exist after some UI steps is in this PR itself. I think adding a try catch block to destroy the VM in any case could be the solution.

**Comment by kbidarkar** [comment]
Just FYI, https://github.com/SatelliteQE/robottelo/pull/2694 
As informed earlier,
I removed [insights] and it's content and moved it to [clients] , the idea is that it's client specific and can be used for both "puppet" and "rhai" , and hence the change. 

Also updated ui/rhai.py

**Comment by kbidarkar** [comment]
@anarang , I think this is what you are looking for, https://github.com/SatelliteQE/robottelo/pull/2694/files#diff-67d0f4984fdd6833ced6dbd2019df14cR511

I think you are facing this issue because 
1) Currently for rhai, the vm context is being created inside `register_client_to_rhai`, we need to avoid this, you will have more control over your testcase if it(vm context) exists under the testcase itself.
2) As suggested by @elyezer , we need to move stuff related to rhai to vm.py like `configure_puppet` or `configure_rhel_repo`. May be plan something like `configure_rhai`
3) I think once you have something like `configure_rhai` in vm module, that would solve your issues.

```
def configure_rhai()
        # Install redhat-access-insights package
        package_name = 'redhat-access-insights'
        result = self.vm.run('yum install -y {0}'.format(package_name))
        if result.return_code != 0:
            raise AccessInsightsError(
                'Unable to install redhat-access-insights rpm'
            )

        # Verify if package is installed by query it
        result = self.vm.run('rpm -q {0}'.format(package_name))
        if result.return_code != 0:
            raise AccessInsightsError(
                'Unable to install redhat-access-insights rpm'
            )

        # Register client with Red Hat Access Insights
        result = self.vm.run('redhat-access-insights --register')
        if result.return_code != 0:
            raise AccessInsightsError(
                'Unable to register client to Access Insights through '
                'Satellite')
```

Hope that helps you.

**Comment by kbidarkar** [comment]
OK, I think you are using register_client_to_rhai to probably save time by using only one VM for all your tests. May be then for tests that require to run a command you could provision another vm?

**Comment by anarang** [comment]
@kbidarkar so, as per the suggestion of adding configure_rhai() to vm.py, the workflow shall be as follows:
```
def test_unregister();
  vm.configure_rhai()
  with Session as session:
    #perform certain clicks on the UI, say, unregister a system form the UI
  result = vm.run('redhat-access-insights')
  # the above step will include going back to the VM and checking if the VM is unregistered or not by excuting the command.
  self.assertEqual(result.return_code, 1)
  # So, the question is if the the above assertion fails then the VM will not be destroyed in the next    step.
  vm.destroy()
```
I have updated the PR with try and catch/finally block to make sure that the VM created is destroyed. I don't mind adding the method to vm.py but the current issue does not seem to be resolved without a try/finally block. Kindly suggest.

**Comment by anarang** [comment]
@elyezer @oshtaier updated the PR with the mentioned comments and handled the vm.destroy() with try/finally block.

**Comment by Ichimonji10** [comment]
@anarang, I see that this idiom is used in two of the three tests:

```python
    def test_something(self):
        try:
            self.rhai.register_client_to_rhai(
                self.ak_name,
                self.org_label,
            )
            # unique test logic
        finally:
            self.rhai.vm.destroy()
```

Why not refactor those statements out in to `setUp` and `tearDown` methods, as
others have suggested?

```python
    def setUp(self):
        self.rhai.register_client_to_rhai(self.ak_name, self.org_label)

    def test_something(self):
        # unique test logic

    def tearDown(self):
        self.rhai.vm.destroy()
```

Place the two tests that require a VM set-up and tear-down in one class and all
other tests in a second class. I notice that there's a pretty intense
`setUpClass` method defined. Move that logic in to a `setUpModule` function.
It'll be structured like this:

```python
def setUpModule():
    pass  # create organization with "insights_" prefix, etc.

class VMTestCase(UITestCase):
    pass  # define setUp, test methods and tearDown

class GenericTestCase(UITestCase):
    pass  # define test methods
```

You can learn more about these methods in the Python standard library documentation for the [`unittest`](https://docs.python.org/3/library/unittest.html) module.

**Comment by Ichimonji10** [comment]
If the current code works (please post output), we can merge this PR now and refactor the RHAI test code in a separate PR. One way or another, though, this code should be refactored so that code isn't copied and pasted.

**Comment by anarang** [comment]
@Ichimonji10 yes the current code works and takes care of the fact that VM gets destroyed. You can merge this PR for now and I 'll look at the refactor part next. Also, I am sure as I add more test cases, will surely refactor the code as per the requirement.
```
----------------------------------------------------------------------
Ran 1 test in 293.731s

OK
```

**Comment by Ichimonji10** [comment]
> You can merge this PR for now and I 'll look at the refactor part next.

Please do. I'll admit that I'm pushing you to [refactor mercilessly](http://www.c2.com/cgi/wiki?RefactorMercilessly) here instead of taking the more lax "[three strikes and you refactor](http://www.c2.com/cgi/wiki?ThreeStrikesAndYouRefactor)" approach. Why? Because, in my opinion, the VM set-up/tear-down code is highly likely to be repeated in more locations and in exactly the same way. I'm making a subjective judgment call about the correct [duplication refactoring threshold](http://www.c2.com/cgi/wiki?DuplicationRefactoringThreshold).

> Also, I am sure as I add more test cases, will surely refactor the code as per the requirement.

I do agree that code should be refactored as more tests are added. Refactoring is an on-going process, not a one-time action, and failing to refactor leads to maintenance nightmares.

**Comment by Ichimonji10** [comment]
ACK

**Comment by elyezer** [comment]
The code now is better and should not leave any running VM. ACK

**Pull request merged**